src/proto/media/file.rs

```diff
@@ -226,3 +433,385 @@ bitflags! {
+            return Err(FileInfoCreationError::InsufficientStorage(info_size));
+        }
+
+        // Write the header at the beginning of the storage
+        let header_ptr = storage.as_mut_ptr() as *mut Header;
+        unsafe {
+            header_ptr.write(header);
+        }
+
+        // At this point, our storage contains a correct header, followed by
+        // random rubbish. It is okay to reinterpret the rubbish as Char16s
+        // because 1/we are going to overwrite it and 2/Char16 does not have a
+        // Drop implementation. Thus, we are now ready to build a correctly
+        // sized &mut Self and go back to the realm of safe code.
+        debug_assert!(!mem::needs_drop::<Char16>());
+        let info_ptr = unsafe {
+            slice::from_raw_parts_mut(storage.as_mut_ptr() as *mut Char16, name_length_ucs2)
+                as *mut [Char16] as *mut Self
+        };
+        let info = unsafe { &mut *info_ptr };
+        debug_assert_eq!(info.name.len(), name_length_ucs2);
+
+        // Write down the UCS-2 name before returning the storage reference
+        for (target, ch) in info.name.iter_mut().zip(name.chars()) {
+            *target = ch
+                .try_into()
+                .map_err(|_| FileInfoCreationError::InvalidChar(ch))?;
+        }
+        info.name[name_length_ucs2 - 1] = NUL_16;
+        Ok(info)
+    }
+}
+
+unsafe impl<Header: FileProtocolInfoHeader> Identify for NamedFileProtocolInfo<Header> {
+    const GUID: Guid = Header::GUID;
+}
+
+impl<Header: FileProtocolInfoHeader> FileProtocolInfo for NamedFileProtocolInfo<Header> {
+    #[allow(clippy::cast_ptr_alignment)]
+    unsafe fn from_uefi<'a>(raw_ptr: *mut c_void) -> &'a mut Self {
+        let byte_ptr = raw_ptr as *mut u8;
+        let name_ptr = byte_ptr.add(mem::size_of::<Header>()) as *mut Char16;
+        let name = CStr16::from_ptr(name_ptr);
+        let name_len = name.to_u16_slice_with_nul().len();
+        let fat_ptr = slice::from_raw_parts_mut(raw_ptr as *mut Char16, name_len);
+        let self_ptr = fat_ptr as *mut [Char16] as *mut Self;
+        &mut *self_ptr
+    }
+}
+
+/// Errors that can occur when creating a FileProtocolInfo
+pub enum FileInfoCreationError {
+    /// The provided buffer was too small to hold the FileInfo. You need at
+    /// least the indicated buffer size (in bytes). Please remember that using
+    /// a misaligned buffer will cause a decrease of usable storage capacity.
+    InsufficientStorage(usize),
+
+    /// The suggested file name contains invalid code points (not in UCS-2)
+    InvalidChar(char),
+}
+
+/// Header for generic file information
+#[repr(C)]
+pub struct FileInfoHeader {
+    size: u64,
+    file_size: u64,
+    physical_size: u64,
+    create_time: Time,
+    last_access_time: Time,
+    modification_time: Time,
+    attribute: FileAttribute,
+}
+
+unsafe impl FileProtocolInfoHeader for FileInfoHeader {
+    const GUID: Guid = Guid::from_values(
```

Comment:
Wait. After looking at the code further, I'm not sure if this is a good idea.

`impl_proto!` is a thing because there are many things we want to do when it comes to protocols:

- Implement Identify in a particular way
    * Tell rustdoc not to expose the Identify impl
    * Tell clippy to ignore the human-readability of GUIDs
- Implement Protocol
- Mark as !Send and !Sync

In exchange for deduplicating all these operations, we are ready to bear with the clunkiness of generating a trait implementation with a macro, which makes the codebase feel more exotic to newcomers, and is most definitely not ideal as soon as lifetimes are involved.

In contrast, only implementing Identify is much simpler, especially if we give up on hiding the GUID from rustdoc and trying to silence a clippy lint that didn't even fire in my experiments. So I'm not sure if the clunkiness of a macro-based trait implementation is still worthwhile in this case.

Reply:
I'm not exactly a fan of lots of macros either. I'm wondering how hard would it be (and if it would be worth it) to get something working with custom derives like `#[derive(Identify("guid"))]`

Reply:
Opened #71 to track this.

Reply:
Should I assume from you creating a separate tracking issue that this part should probably be left out of this PR?

Reply:
Well, this PR is already making a lot of changes, and macros would make it even harder to review. So it would be great to get this through first.